### PR TITLE
expression: implement vectorized evaluation for `builtinLogicAndSig`

### PR DESCRIPTION
**Pull request by lauhg**
### What problem does this PR solve? 
implement vectorized evaluation for `builtinLogicAndSig`
Issue: #12103

### What is changed and how it works?
```sh
BenchmarkVectorizedBuiltinOpFunc/builtinLogicAndSig-VecBuiltinFunc-4         	  218979	      5192 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedBuiltinOpFunc/builtinLogicAndSig-NonVecBuiltinFunc-4      	   33757	     33742 ns/op	       0 B/op	       0 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_op_vec.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+func (b *builtinLogicAndSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+
+	buf0, err := b.bufAllocator.get(types.ETInt, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf0)
+	if err := b.args[0].VecEvalInt(b.ctx, input, buf0); err != nil {
```
Comment (Reminiscent) [inline expression/builtin_op_vec.go]: We can directly use the result variable to store the result of args[0]. Thus we can improve the performance by saving a memory copy.

File: expression/builtin_op_vec.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[1].VecEvalInt(b.ctx, input, buf1); err != nil {
+		return err
+	}
+
+	i64s := result.Int64s()
+	arg1 := buf1.Int64s()
```
Comment (zz-jason) [inline expression/builtin_op_vec.go]: You can use `result.MergeNulls(buf1)` to calculate the null bitmap for `result`. It's much faster.

Comment (lauhg) [reply expression/builtin_op_vec.go]:
```go
isNull0 := result.IsNull(i)
if !isNull0 && i64s[i] == 0 {
	i64s[i] = 0
	continue
}
```
@zz-jason  Here will check if resut isNull, I think use `result.MergeNulls(buf1)` may cause bugs.

File: expression/builtin_op_vec.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		if !isNull1 && arg1[i] == 0 {
+			i64s[i] = 0
+			continue
+		}
+
+		if isNull0 || isNull1 {
+			result.SetNull(i, true)
+			continue
+		}
+
```
Comment (Reminiscent) [inline expression/builtin_op_vec.go]: The problem is that if arg0 = null , arg1 = 0. then ```result.isNull(i) = null```. But we need ```result.isNull(i) = false```.

**Comment by Reminiscent** [comment]
I think [this](https://github.com/pingcap/tidb/pull/12365/files#diff-c02f54138c0c06b976278e91d3832d2aR39-R56) kind of implementation will be more elegant, what's your opinion? @zz-jason  @qw4990 @lauhg

**Comment by lauhg** [comment]
> I think [this](https://github.com/pingcap/tidb/pull/12365/files#diff-c02f54138c0c06b976278e91d3832d2aR39-R56) kind of implementation will be more elegant, what's your opinion? @zz-jason @qw4990 @lauhg

But this will introduce unnecessary computations.

**Comment by qw4990** [comment]
@lauhg Please solve the conflict~

**Pull request merged**